content/en/docs/setup/version-skew-policy.md

```diff
@@ -0,0 +1,127 @@
+
+### `kube-controller-manager` and `kube-scheduler`
+
+`kube-controller-manager` and `kube-scheduler` must not be newer than the `kube-apiserver` instances they communicate with. They are expected to match the `kube-apiserver` minor version, but may be up to one minor version older (to allow live upgrades).
+
+Example:
+
+* `kube-apiserver` is at `1.13`
+* `kube-controller-manager` and `kube-scheduler` are supported at `1.13` and `1.12`
+
+Note that if version skew exists between `kube-apiserver` instances in a multi-server cluster, and `kube-controller-manager` or `kube-scheduler` can communicate with any `kube-apiserver` instance in the cluster (for example, via a load balancer), this narrows the allowed `kube-controller-manager` or `kube-scheduler` versions.
+
+Example:
+
+* `kube-apiserver` instances are at `1.13` and `1.12`
+* `kube-controller-manager` and `kube-scheduler` communicate with a load balancer that can route to any `kube-apiserver` instance
+* `kube-controller-manager` and `kube-scheduler` are supported at `1.12` (`1.13` is not supported because that would be newer than the `kube-apiserver` instance at version `1.12`)
+
+### `kubectl`
+
+`kubectl` is supported within one minor version (older or newer) of `kube-apiserver`.
+
+Example:
+
+* `kube-apiserver` is at `1.13`
+* `kubectl` is supported at `1.14`, `1.13`, and `1.12`
+
+Note that if version skew exists between `kube-apiserver` instances in a multi-server cluster, this narrows the supported `kubectl` versions.
+
+Example:
+
+* `kube-apiserver` instances are at `1.13` and `1.12`
+* `kubectl` is supported at `1.13` and `1.12` (other versions would be more than one minor version skewed from one of the `kube-apiserver` components)
+
+## Supported component upgrade order
+
+The supported version skew between components has implications on the order in which components must be upgraded.
+This section describes the order in which components must be upgraded to transition an existing cluster from version `1.n` to version `1.(n+1)`.
+
+### `kube-apiserver`
+
+Pre-requisites:
+
+* In a multi-server cluster, other `kube-apiserver` instances are at `1.n` or `1.(n+1)` (this ensures maximum skew of 1 minor version between the oldest and newest `kube-apiserver` instance)
+* The `kube-controller-manager` and `kube-scheduler` instances that communicate with this server are at version `1.n` (this ensures they are not newer than the existing API server version, and are within 1 minor version of the new API server version)
+* `kubelet` instances on all nodes are at version `1.n` or `1.(n-1)` (this ensures they are not newer than the existing API server version, and are within 2 minor versions of the new API server version)
+* Registered admission webhooks are capable of understanding all the versions of REST resources they are registered for (including any new versions available in `1.(n+1)`)
+
+Upgrade `kube-apiserver` to `1.(n+1)`
+
+### `kube-controller-manager` and `kube-scheduler`
```

Comment:
Do you intentionally leave out cloud controller manager? What about supported skew between arbitrary controllers and the apiserver (e.g. ingress, autoscaler, etc) in terms of their client-go and/or controller-runtime packages? If we say that we only support 1 version skew for kubectl (a client) but two for kubelet (a different client), what is the supported skew for an arbitrary client that links in our official client libraries?

Reply:
> Do you intentionally leave out cloud controller manager?

I wasn't sure what the state of that controller was. The [docs](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) still look to require alpha features (initializers) to be enabled to get parity with kube-controller-manager. If that is supported without using alpha features, we should probably include it.

> What about supported skew between arbitrary controllers and the apiserver (e.g. ingress, autoscaler, etc) in terms of their client-go and/or controller-runtime packages?

I think we would have to defer to those components for what skew to the apiserver they support.

> If we say that we only support 1 version skew for kubectl (a client) but two for kubelet (a different client), what is the supported skew for an arbitrary client that links in our official client libraries?

It completely depends on how they use the libraries. For example, kubectl uses unstructured clients for basic CRUD so it is forward and backward compatible. It also waits to use typed clients for new API versions for one release so that it can run against an API server at the previous version.